src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -65,52 +67,272 @@ ucs_status_t uct_rdmacm_cm_reject(struct rdma_cm_id *id)
+    return UCS_ERR_UNSUPPORTED;
+}
+
+static ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_list_free(struct uct_rdmacm_cm_qp_context *ctx)
+{
+    ucs_status_t status = UCS_OK;
+
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk, *tmp;
+
+    pthread_mutex_lock(&ctx->reserved_qpn_ctx.mutex);
+
+    ucs_list_for_each_safe(blk, tmp, &ctx->reserved_qpn_ctx.blk_list, entry) {
+        ucs_list_del(&blk->entry);
+        status = uct_rdmacm_cm_reserved_qpn_blk_free(blk);
+        if (status != UCS_OK) {
+            ucs_error("uct_rdmacm_cm_reserved_qpn_blk_list_free failed");
+        }
+    }
+
+    pthread_mutex_unlock(&ctx->reserved_qpn_ctx.mutex);
+#endif
+
+    return status;
+}
+
+static void uct_rdmacm_cm_qp_context_cleanup(uct_rdmacm_cm_t *cm)
+{
+    struct uct_rdmacm_cm_qp_context *ctx;
+    ucs_status_t status;
+    int ret;
+
+    kh_foreach_value(&cm->ctxs, ctx, {
+        if (ctx->reserved_qpn_support) {
+            /* There can be some blks are not fully used, then they won't be 
+               destroyed in RDMACM CM EP, so need to be destroyed here. */
+            status = uct_rdmacm_cm_reserved_qpn_blk_list_free(ctx);
+            ucs_assert_always(status == UCS_OK);
+        } else {
+            ret = ibv_destroy_cq(ctx->dummy_qp_ctx.cq);
+            if (ret != 0) {
+                ucs_warn("ibv_destroy_cq() returned %d: %m", ret);
+            }
+        }
+        
+        ucs_free(ctx);
+    });
+
+    kh_destroy_inplace(uct_rdmacm_cm_qp_contexts, &cm->ctxs);
+}
+
+ucs_status_t uct_rdmacm_cm_get_qp_context(uct_rdmacm_cm_t *cm,
+                                          struct ibv_context *verbs,
+                                          struct uct_rdmacm_cm_qp_context **ctx_p)
 {
-    struct ibv_cq *cq;
+    struct uct_rdmacm_cm_qp_context *ctx;
+    ucs_status_t status;
     khiter_t iter;
     int ret;
 
-    iter = kh_put(uct_rdmacm_cm_cqs, &cm->cqs,
+    iter = kh_put(uct_rdmacm_cm_qp_contexts, &cm->ctxs,
                   ibv_get_device_guid(verbs->device), &ret);
     if (ret == -1) {
-        ucs_error("cm %p: cannot allocate hash entry for CQ", cm);
-        return UCS_ERR_NO_MEMORY;
+        ucs_error("cm %p: cannot allocate hash entry for qp context", cm);
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
     }
 
     if (ret == 0) {
         /* already exists so use it */
-        cq = kh_value(&cm->cqs, iter);
+        ctx = kh_value(&cm->ctxs, iter);
     } else {
-        /* Create a dummy completion queue */
-        cq = ibv_create_cq(verbs, 1, NULL, NULL, 0);
-        if (cq == NULL) {
-            kh_del(uct_rdmacm_cm_cqs, &cm->cqs, iter);
-            ucs_error("ibv_create_cq() failed: %m");
-            return UCS_ERR_IO_ERROR;
+        /* Create a qp context */
+        ctx = ucs_malloc(sizeof(struct uct_rdmacm_cm_qp_context),
+                         "qp_context");
+        if (ctx == NULL) {
+            ucs_error("cm %p: malloc qp context failed: %m", cm);
```

Comment:
`ucs_error("cm %p: failed to allocate device context", cm);`

Reply:
@yosefe Fixed.